plugin/server/datastore-sqlite/sqlite.go

```diff
@@ -520,6 +520,75 @@ func (sqlitePlugin) ListSpiffeEntries(
 	return &datastore.ListSpiffeEntriesResponse{}, errors.New("Not Implemented")
 }
 
+// RegisterToken takes a Token message and stores it
+func (ds *sqlitePlugin) RegisterToken(req *datastore.JoinToken) (*common.Empty, error) {
+	var resp *common.Empty
+	if req.Token == "" || req.Expiry == 0 {
+		return resp, errors.New("token and expiry are required")
+	}
+
+	t := joinToken{
+		Token:  req.Token,
+		Expiry: req.Expiry,
+	}
+
+	return resp, ds.db.Create(&t).Error
+}
+
+// FetchToken takes a Token message and returns one, populating the fields
+// we have knowledge of
+func (ds *sqlitePlugin) FetchToken(req *datastore.JoinToken) (*datastore.JoinToken, error) {
+	var t joinToken
+
+	err := ds.db.Find(&t, "token = ?", req.Token).Error
+	if err == gorm.ErrRecordNotFound {
+		return &datastore.JoinToken{}, nil
+	}
+
+	resp := &datastore.JoinToken{
+		Token:  t.Token,
+		Expiry: t.Expiry,
+	}
+	return resp, err
+}
+
+func (ds *sqlitePlugin) DeleteToken(req *datastore.JoinToken) (*common.Empty, error) {
+	var resp *common.Empty
+
+	// Protect the data - if gorm gets a delete w/ an empty primary
+	// key, it deletes _all_ the records...
+	if req.Token == "" {
+		return &common.Empty{}, errors.New("no token specified")
+	}
+
+	t := joinToken{
+		Token:  req.Token,
+		Expiry: req.Expiry,
+	}
+	return resp, ds.db.Delete(&t).Error
```

Comment:
soft delete vs a hard delete? Do we see a use case where we might want to persist all the join tokens generated.

Reply:
Current implementation uses soft delete magic provided by gorm.

I actually didn't really want to do soft delete here because I could see this running away fairly easily on a large deployment, however using soft delete seemed to be the gorm happy path so I took it

Reply:
If you don't have a `deleted_at` field it'll do hard deletes.

Reply:
Hmm interesting... Originally, I tried anonymous import on model of gorm.Model, and omitted DeletedAt column in the schema (since the docs did make it seem like this was possible), but I received an error about the missing column when I tried...

Maybe if I define CreatedAt and UpdatedAt manually it will work

Reply:
yeah, weird.  IIRC `DeletedAt` was a pointer so that it could not be defined.  I've had similar issues & frustrations w/ gorm before tho so /shrug.